toggl-open-source/toggldesktop
# Make Autotracker support time range and days of week

### 📒 Description
Make Autotracker support time range and days of week

### 🕶️ Types of changes
- **New feature** (non-breaking change which adds functionality)
- **Breaking change** (fix or feature that would cause existing functionality to change)

### 🤯 List of changes
- [x] Saving to / reading from the database
- [x] Matching a time range and days of the week with timeline events
- [x] Converting into view struct for passing to the platform clients
- [x] Change the API method for adding an autotracker rule
- [x] Change the API method for editing an autotracker rule
- [x] Make sure backward compatibility is maintained (add DB migration if necessary)
- [x] App tests
- [x] API tests

### 👫 Relationships
Closes #3916
Closes #3855 

### 🔎 Review hints

The code review should be the main focus, as this PR doesn't change any user-facing functionality.
The thing I'm most unsure of is the days of week API - I decided to go for `uint8` as a representation of days of week bitset for the reason of API simplicity (1 field instead of 7 fields or a separate struct) and memory efficiency. The downside is that it's not clear which bit corresponds to which days of the week. This is why I added a small API method `toggl_days_of_week_into_uint8()` that can be used to make the conversion or as a reference.

Manual test scenario:
1. Run the app to make sure the database has been migrated. Verify the app is running without issues.
2. Create an autotracker rule and save it.
3. Open the database file and set start time, end time, and/or days of the week to the added autotracker rule.
4. Reopen the app to make sure the edited rule is loaded.
5. See if the start time, end time, and days of the week settings are respected.
